### PR TITLE
[feature] configuration validation

### PR DESCRIPTION
**Pull request by jbottigliero**
- uses JSON schema to validate configuration types.
  - validated via `webpack/lib/validateSchema`
- adds tests for various configuration types
  - adds negative tests for an invalid type and array.

**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
Yes – I'll probably add more based on actual desired implementation.

**If relevant, did you update the documentation?**
N/A – for now.

**Summary**

This is an port of webpack/webpack#6255 which is intended to resolve webpack/webpack#4977.

**Does this PR introduce a breaking change?**
> No, it's not expected to... However, it may cause previously invalid configuration objects to throw errors not seen before.

**Other information**

Some outstanding questions (from webpack/webpack#6255):

> - Is using JSON schema validation too much for this?
>   - It does seem like it might be better to combine this validation with the configuration options validation. This would be a much larger change and is worthy of a discussion.
> - Should it be using the `WebpackOptionsValidationError` class? The updated output still leaves a bit to be desired. This could be accomplished using the same class or it's own.

With the code now being in separate repositories/modules, I could see how the proposed solution is less than ideal. I'm up for discussing how to move forward!

### CODE REVIEW

**Comment by jbottigliero** [comment]
I _believe_ the failing tests here are addressed in #238

**Comment by ematipico** [comment]
Yeah exactly, these are tests due to the new version of `webpack` and how it outputs the entry points.

**Comment by ev1stensberg** [comment]
Let's wait for the fix on that to land and then do a review/rebase

File: bin/convert-argv.js [hunk 1]
```diff
@@ -3,6 +3,10 @@ var fs = require("fs");
 fs.existsSync = fs.existsSync || path.existsSync;
 var interpret = require("interpret");
 var prepareOptions = require("./prepareOptions");
+var webpackConfigurationSchema = require("../schemas/webpackConfigurationSchema.json");
+var validateSchema = require("webpack/lib/validateSchema");
+var WebpackOptionsValidationError = require("webpack/lib/WebpackOptionsValidationError");
```
Comment (ooflorent) [inline bin/convert-argv.js]: Please use `webpack.WebpackOptionsValidationError` instead

File: bin/convert-argv.js [hunk 2]
```diff
@@ -3,6 +3,10 @@ var fs = require("fs");
 fs.existsSync = fs.existsSync || path.existsSync;
 var interpret = require("interpret");
 var prepareOptions = require("./prepareOptions");
+var webpackConfigurationSchema = require("../schemas/webpackConfigurationSchema.json");
+var validateSchema = require("webpack/lib/validateSchema");
```
Comment (ooflorent) [inline bin/convert-argv.js]: Please use `webpack.validateSchema ` instead

**Comment by ev1stensberg** [comment]
@jbottigliero mind rebasing against the master branch?

**Comment by jbottigliero** [comment]
@ev1stensberg – rebased and updated. I wasn't able to reproduce the test failure in `node@6` locally, should I just re-trigger the build? Any insight here would be helpful!

@ooflorent – for the usage of `webpack.*`... I was basing my implementation/usage of webpack lib files off of `lib/commands/migration.js`. I'm not super familiar with the build process and this package's association between `webpack` core, yet. Is it safe to require webpack core here? I'll dig through the code myself, but any advice in direction would be helpful!

**Comment by ev1stensberg** [comment]
@jbottigliero sure, happens sometimes, let's do another try

**Comment by ev1stensberg** [comment]
Re-started your build, let's cross fingers 🤞

File: bin/convert-argv.js (hunk 2 above)
Comment (TheLarkInn) [reply bin/convert-argv.js]: Can you switch this over @jbottigliero?

**Comment by jbottigliero** [comment]
> @ooflorent – for the usage of `webpack.*`... I was basing my implementation/usage of webpack lib files off of `lib/commands/migration.js`. I'm not super familiar with the build process and this package's association between `webpack` core, yet. Is it safe to require webpack core here? I'll dig through the code myself, but any advice in direction would be helpful!

Updated based on #229.

cc. @TheLarkInn

**Comment by jbottigliero** [comment]
Thanks for taking a look @ev1stensberg and @fokusferit.

After taking another look at the code in `bin/convert-argv.js`, maybe the validation could be more strict? 

When `processConfiguredOptions` [is called](https://github.com/webpack/webpack-cli/blob/master/bin/convert-argv.js#L141), it looks like the configuration can only ever be a `Promise`, object or array of objects.

When you're using a `function` for a configuration, it's [_actually_ handled by `prepareOptions`](https://github.com/webpack/webpack-cli/blob/master/bin/convert-argv.js#L130).

Seems like I could/should also add a case for the `Promise` configuration.

**Comment by fokusferit** [comment]
Hey @jbottigliero what do you think regarding about adding a test case for `Promise`configuration? Just wanted to follow up :)

**Comment by ev1stensberg** [comment]
If we can strip the whole function and add a Promise clause instead that should be preferable

**Comment by jbottigliero** [comment]
@fokusferit, @ev1stensberg, just took a look at this...

Unless I'm mistaking there are few things that need to happen in order to make this work (at least to validate the `Promise` instance).

It appears the `ajv-keywords` package [doesn't actually support `"instanceof": "Promise"`](https://github.com/epoberezkin/ajv-keywords/blob/master/keywords/instanceof.js#L3). I plan to open a ticket there to discuss whether or not it could/should be added – I'll follow-up on this.

An alternative would be to add a custom validator similar to[`webpack`'s addition of `absolutePath`
](https://github.com/webpack/webpack/blob/master/schemas/ajv.absolutePath.js)... because we're sharing the validator, I'll need to open a PR against `webpack` core.

I think ideally we get `Promise` added to the `ajv-keywords` package.

For now, I've added two new tests and resolved merge conflicts.

**Comment by jbottigliero** [comment]
(tests are passing now because it's still just `"instanceof": "Function"`)

**Comment by jbottigliero** [comment]
Support for `"instanceof": "Promise"` was just merged – epoberezkin/ajv-keywords#51 (closed epoberezkin/ajv-keywords#51).

Once released, I'll open a PR bumping `webpack` core's dependency version and update here.

**Comment by ev1stensberg** [comment]
Maybe we can keep it like this an open an issue up for discussion? Instead of rushing it

**Comment by jbottigliero** [comment]
@ev1stensberg works for me. I _did_ just open webpack/webpack#6430...

The dependency version bump in core _does_ feel pretty heavy-handed based on the original intent of this PR.

I'll be watching for feedback, thanks again.

**Comment by ev1stensberg** [comment]
Can you make an issue at webpack-cli with what we've just discussed and link to the  6430 and this pr? Thanks!

**Comment by jbottigliero** [comment]
The [console.error](https://github.com/webpack/webpack-cli/pull/240/files#diff-daabbe74e88640829a3ad9e9899efd8cR163) will log:

```
Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration should be one of these:
   object | [object] | function
   Details:
    * configuration should be an object.
      -> A webpack configuration object.
    * configuration should be an array:
      [object]
      -> An array of webpack configuration objects.
    * configuration should be an instance of function
      -> A promise that resolves with a configuration object, or an array of configuration objects.
```

I can certainly add something along the lines of `Received: ${typeof options} (${options})` to this.

**Comment by ev1stensberg** [comment]
Previous looked fine, but some extra indication would be nice, like the schema, but:

```

Webpack has been initialised using a configuration object that does not match the API schema. Recieved: $type 
...
```

**Comment by ev1stensberg** [comment]
But that might be a schema thing, so LGTM :shipit:

File: bin/convert-argv.js [hunk 3]
```diff
@@ -148,9 +152,17 @@ module.exports = function(...args) {
...
+			options
+		);
+		if (webpackConfigurationValidationErrors.length) {
+			var error = new WebpackOptionsValidationError(
+				webpackConfigurationValidationErrors
+			);
 			console.error(
-				"Config did not export an object or a function returning an object."
+				error.message,
+				`\nReceived: ${typeof options} : ${JSON.stringify(options, null, 2)}`
```
Comment (ev1stensberg) [inline bin/convert-argv.js]: Check the json stringify callback to b correct @fokusferit @ematipico

Comment (ematipico) [reply bin/convert-argv.js]: I would add a newline `\n` straight after `:`. But the rest looks good to me

**Comment by ematipico** [comment]
I really like the correct indentation of the configuration error, it makes more cleaner and easier to read!

**Comment by ev1stensberg** [comment]
Thanks ☝️💙💙

**Pull request merged**